### PR TITLE
Feature/shared pointer

### PR DESCRIPTION
**Pull request by artem-zyktin**
- Add SharedPtr and tests for it
- Fix include paths in CoreLibsTest project

### CODE REVIEW

File: CoreLib/SharedPtr.h [hunk 1]
```diff
@@ -0,0 +1,490 @@
+﻿#pragma once
+
+
+#include "CoreLibConcepts.h"
+
+#include <Core.h>
+
+#include <stdint.h>
```
Comment (LtSnail) [inline CoreLib/SharedPtr.h:8]: Nitpick: Maybe we can just include it to pch?

File: CoreLib/SharedPtr.h [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+        void Reset(std::remove_extent_t<T>* ptr = nullptr);
+        void Release();
+
+        void Swap(SharedPtr<T, Deleter>& other) noexcept;
+
+        template<class U, class DeleterU> requires IsDefaultCompatibleDeleter<U, DeleterU>
+        friend class WeakPtr;
+
+        template<class U>
+        bool operator==(std::remove_extent_t<U>* other) const noexcept;
```
Comment (LtSnail) [inline CoreLib/SharedPtr.h:78]: https://en.cppreference.com/w/cpp/memory/shared_ptr/operator_cmp

Does it make sense to follow the pre C++20 API here?

File: CoreLib/SharedPtr.h [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+        _control_block = nullptr;
+    }
+
+    template<class T, class Deleter> requires IsDefaultCompatibleDeleter<T, Deleter>
+    inline void WeakPtr<T, Deleter>::Swap(WeakPtr<T, Deleter>& other) noexcept
+    {
+        std::swap(_control_block, other._control_block);
+    }
+
+
```
Comment (LtSnail) [inline CoreLib/SharedPtr.h:232]: Too much blank lines

File: CoreLib/SharedPtr.h [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+    template<class T, class Deleter> requires IsDefaultCompatibleDeleter<T, Deleter>
+    class SharedPtr;
+
+    template<class T, class Deleter = std::default_delete<T>> requires IsDefaultCompatibleDeleter<T, Deleter>
+    class FUEGO_API WeakPtr
+    {
+    public:
+
+        WeakPtr() noexcept = default;
+        WeakPtr(const WeakPtr<T, Deleter>& other) noexcept;
```
Comment (LtSnail) [inline CoreLib/SharedPtr.h:21]: What do you think about `using` for `WeakPtr<T, Deleter>` and `SharedPtr<T, Deleter>`?

**Review by LtSnail — APPROVED** [review]

File: CoreLib/SharedPtr.h (hunk 2 above)
Comment (artem-zyktin) [reply CoreLib/SharedPtr.h:78]: Good point

File: CoreLib/SharedPtr.h (hunk 3 above)
Comment (artem-zyktin) [reply CoreLib/SharedPtr.h:232]: These are different blocks - division between blocks WeakPtr in SharedPtr

Comment (LtSnail) [reply CoreLib/SharedPtr.h:232]: It's better to mark it with a comment then

**Pull request merged**